x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/action/TestGrokPatternAction.java

```diff
@@ -0,0 +1,194 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.grok.GrokCaptureExtracter;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+public class TestGrokPatternAction extends ActionType<TestGrokPatternAction.Response> {
+
+    public static final TestGrokPatternAction INSTANCE = new TestGrokPatternAction();
+    public static final String NAME = "cluster:monitor/text_structure/test_grok_pattern";
+
+    private TestGrokPatternAction() {
+        super(NAME, Response::new);
+    }
+
+    public static class Request extends ActionRequest {
+
+        public static final ParseField GROK_PATTERN = new ParseField("grok_pattern");
+        public static final ParseField TEXT = new ParseField("text");
+
+        private static final ObjectParser<Request.Builder, Void> PARSER = createParser();
+
+        private static ObjectParser<Request.Builder, Void> createParser() {
+            ObjectParser<Request.Builder, Void> parser = new ObjectParser<>("textstructure/testgrokpattern", false, Request.Builder::new);
+            parser.declareString(Request.Builder::grokPattern, GROK_PATTERN);
+            parser.declareStringArray(Request.Builder::text, TEXT);
+            return parser;
+        }
+
+        public static class Builder {
+            private String grokPattern;
+            private List<String> text;
+
+            public Builder grokPattern(String grokPattern) {
+                this.grokPattern = grokPattern;
+                return this;
+            }
+
+            public Builder text(List<String> text) {
+                this.text = text;
+                return this;
+            }
+
+            public Request build() {
+                return new Request(grokPattern, text);
+            }
+        }
+
+        private final String grokPattern;
+        private final List<String> text;
+
+        private Request(String grokPattern, List<String> text) {
+            this.grokPattern = grokPattern;
+            this.text = text;
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            grokPattern = in.readString();
+            text = in.readStringCollectionAsList();
+        }
+
+        public static Request parseRequest(XContentParser parser) throws IOException {
+            return PARSER.parse(parser, null).build();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(grokPattern);
```

Comment:
This line would throw an obscure exception if `grokPattern` was `null`, which is why it's best to prevent that in the constructor.